RecoTracker/MkFitCMS/src/MkStdSeqs.cc

```diff
@@ -586,39 +594,131 @@ namespace mkfit {
+                                                      clean_duplicates_sharedhits_pixelseed);
+        }
+      } rdc_instance;
+    }  // namespace
+
     //=========================================================================
-    //
+    // Quality filters
     //=========================================================================
 
-    void find_and_remove_duplicates(TrackVec &tracks, const IterationConfig &itconf) {
-#ifdef DEBUG
-      std::cout << " find_and_remove_duplicates: input track size " << tracks.size() << std::endl;
-#endif
-      if (itconf.m_requires_quality_filter && !(itconf.m_requires_dupclean_tight)) {
-        find_duplicates_sharedhits(tracks, itconf.m_params.fracSharedHits);
-      } else if (itconf.m_requires_dupclean_tight) {
-        find_duplicates_sharedhits_pixelseed(tracks,
-                                             itconf.m_params.fracSharedHits,
-                                             itconf.m_params.drth_central,
-                                             itconf.m_params.drth_obarrel,
-                                             itconf.m_params.drth_forward);
-      } else {
-        find_duplicates(tracks);
-        remove_duplicates(tracks);
-      }
+    // quality filter for n hits with seed hit "penalty" for strip-based seeds
+    //   this implicitly separates triplets and doublet seeds with glued layers
+    template <class TRACK>
+    bool qfilter_n_hits(const TRACK &t, const MkJob &j) {
+      int seedHits = t.getNSeedHits();
+      int seedReduction = (seedHits <= 5) ? 2 : 3;
+      return t.nFoundHits() - seedReduction >= j.params_cur().minHitsQF;
+    }
 
-#ifdef DEBUG
-      std::cout << " find_and_remove_duplicates: output track size " << tracks.size() << std::endl;
-      for (auto const &tk : tracks) {
-        std::cout << tk.parameters() << std::endl;
-      }
-#endif
+    // simple hit-count quality filter; used with pixel-based seeds
+    template <class TRACK>
+    bool qfilter_n_hits_pixseed(const TRACK &t, const MkJob &j) {
+      return t.nFoundHits() >= j.params_cur().minHitsQF;
+    }
+
+    // layer-dependent quality filter
+    template <class TRACK>
+    bool qfilter_n_layers(const TRACK &t, const MkJob &j) {
+      const BeamSpot &bspot = j.m_beam_spot;
+      const TrackerInfo &trk_inf = j.m_trk_info;
+      int enhits = t.nHitsByTypeEncoded(trk_inf);
+      int npixhits = t.nPixelDecoded(enhits);
+      int enlyrs = t.nLayersByTypeEncoded(trk_inf);
+      int npixlyrs = t.nPixelDecoded(enlyrs);
+      int nmatlyrs = t.nTotMatchDecoded(enlyrs);
+      int llyr = t.getLastFoundHitLyr();
+      int lplyr = t.getLastFoundPixelHitLyr();
+      float invpt = t.invpT();
+      float invptmin = 1.43;  // min 1/pT (=1/0.7) for full filter on (npixhits<=3 .or. npixlyrs<=3)
+      float d0BS = t.d0BeamSpot(bspot.x, bspot.y);
+      float d0_max = 0.1;  // 1 mm
+
+      bool endsInsidePix = (llyr == 2 || llyr == 18 || llyr == 45);
```

Comment:
more magic numbers. Should they go into a config or get pulled from a DB?

Reply:
old code, already included in todo with #36966

there will be no DB because the indices correspond to compiled code; using named constants was a plan for the todo

Reply:
> Should they go into a config or get pulled from a DB?

same (improved response), this is already configurable at the level of a function.